### PR TITLE
HPCC-22971 Avoid creating compound disk reads for alien types

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
@jakesmith @richardkchapman do we agree this is a good idea?  See jira for details of examples that change.

**Comment by jakesmith** [comment]
>If we never created compound activities for datasets with alien data types and complex ifblocks then it would allow a very simple disk activity to be kept for the case where alien datatypes are mixed with virtual fields.

I'm not clear on what you mean here.
If you prevent compound activities being generated in alien/ifblock cases, that do you mean by it "would allow a very simple disk activity to be kept for the case where alien datatypes are mixed with virtual fields" ?

**Comment by ghalliday** [comment]
Response to Jake:
Going forward the fileposition needs to be handled by the translation code (e.g. to support remote reading and filtering), but alien data types are not supported by the translation code - so we need another approach or to disallow the functionality.  But it is legitimate to want to use complex ifs/alien data types and and filepositions when reading external data files..

It would be a pain (if not impossible) to retain the current disk reading code with all its complexity including updating a local fileposition.
However it would be reasonable to have a disk read activity which only supports an absolute minimum of functionality which can be used for this case (and selected with a flag supplied by the code generator).  This allows the general case to be as efficient as possible, and the code for this weird special case to be as minimal as possible.

**Comment by jakesmith** [comment]
Are there real world cases in use, where alien data types and virtual fields are used together?

**Comment by ghalliday** [comment]
Yes there are (e.g. using an alien data type to represent an image which is fetched from a file using fetch).  In that case they should not be using an alien data type, but I think there are also other more legitimate cases when it is an external data file.

File: ecl/hql/hqlattr.cpp [hunk 1]
```diff
@@ -4069,3 +4069,82 @@ CHqlMetaProperty * queryMetaProperty(IHqlExpression * expr)
...
+            case type_row:
+                if (hasReferenceModifier(type))  // Never currently generated
+                    return false;
+                if (!canDefinitelyProcessWithTranslator(cur->queryRecord()))
+                    return false;
+                break;
+            case type_dictionary:
+            case type_groupedtable:
+            case type_table:
+                {
```
Comment (jakesmith) [inline ecl/hql/hqlattr.cpp]: pedantic/formatting: inconsistent indentation of brace vs other case statement (line 4094).

**Comment by jakesmith** [comment]
@ghalliday - looks sensible to me

**Comment by ghalliday** [comment]
@richardkchapman please review

**Pull request merged**